common/src/main/scala/tracking/NotificationReportRepository.scala

```diff
@@ -35,30 +36,48 @@ class NotificationReportRepository(client: AsyncDynamo, tableName: String)
       return Future.successful(Left(RepositoryError("Date range too big to query")))
     }
 
-    def maybeLastKey(result: QueryResult): Option[util.Map[String, AttributeValue]] = Option(result.getLastEvaluatedKey).flatMap(lastKey => if (lastKey.isEmpty) None else Some(lastKey))
+    def maybeStartKey(result: QueryResult): Option[util.Map[String, AttributeValue]] = Option(result.getLastEvaluatedKey).flatMap(lastKey => if (lastKey.isEmpty) None else Some(lastKey))
 
-    def reportsFromResult(result: QueryResult): List[NotificationReport] = result.getItems.asScala.toList.flatMap { item =>
-      fromAttributeMap[NotificationReport](item.asScala.toMap).asOpt
+    def reportsFromResult(result: QueryResult): RepositoryResult[List[NotificationReport]] = {
+      val results = result.getItems.asScala.toList.map { item =>
+        fromAttributeMap[NotificationReport](item.asScala.toMap)
+      }
+      val error = results.collectFirst {
+        case JsError(errors) => Left(RepositoryError(s"Unable to parse notification report $errors"))
+      }
+
+      val reports = results.collect{
+        case JsSuccess(report, _) => report
+      }
+
+      error.getOrElse(Right(reports))
     }
 
-    def query(): QueryRequest = new QueryRequest(tableName)
+    def buildDynamoQuery(startKey: Option[util.Map[String, AttributeValue]]): QueryRequest = new QueryRequest(tableName)
       .withIndexName(SentTimeIndex)
       .withKeyConditions(Map(
         TypeField -> keyEquals(notificationType.value),
         SentTimeField -> keyBetween(from.toString, to.toString)
       ).asJava)
+      .withExclusiveStartKey(startKey.orNull)
 
-    def queryWithKey(lastKey: util.Map[String, AttributeValue]): QueryRequest = query().withExclusiveStartKey(lastKey)
-
-    def enrichWithRest(lastKey: util.Map[String, AttributeValue], lastList: List[NotificationReport]): Future[List[NotificationReport]] = client.query(queryWithKey(lastKey)) flatMap { result =>
-      val reports = lastList ++ reportsFromResult(result)
-      maybeLastKey(result).map(enrichWithRest(_, reports)).getOrElse(Future.successful(reports))
+    def fetch(
+      startKey: Option[util.Map[String, AttributeValue]] = None,
+      lastList: RepositoryResult[List[NotificationReport]] = Right(Nil)
+    ): Future[RepositoryResult[List[NotificationReport]]] = {
+      client.query(buildDynamoQuery(startKey)) flatMap { result =>
+        val reports = for {
+          list1 <- lastList
+          list2 <- reportsFromResult(result)
```

Comment:
Can these be a bit more descriptive? `list1` and `list2` is not that clear at the moment